Feat/gtfs position tracker

1Maxnet1 opened the pull request:
This implements a predictor based on the gtfs-position-tracker. Currently it is implemented and runs, but not yet verified, whether it works or not. However, we might merge it already, so the codebase does not diverge to much and have a base, where multiple people can work on.

Dirtstone committed:
feat: Adding simple predictor

1Maxnet1 committed:
Create the initial structure for gtfs-position-tracker

1Maxnet1 committed:
Outline how to implement gtfs-position-tracker.cc

1Maxnet1 committed:
Retrieve trip ID for vehicle position

1Maxnet1 committed:
Retrieve a list of stops for the given trip ID

1Maxnet1 committed:
Calculate the distance between two points

1Maxnet1 committed:
Check if a tripUpdate already exists

1Maxnet1 committed:
Update existing TripUpdate if it exists

1Maxnet1 committed:
Create new tripUpdate if needed to now instead of

1Maxnet1 committed:
Remove redundant code

1Maxnet1 committed:
Shift includes to the beginning of the file

1Maxnet1 committed:
Fixing build

1Maxnet1 committed:
Make use of gtfs-position-tracker predictor

1Maxnet1 committed:
Fix accessing wrong feed

1Maxnet1 committed:
Add debugging prints

1Maxnet1 committed:
Initialize points

1Maxnet1 committed:
Catching issues with Stop-Index and Trip-ID

1Maxnet1 committed:
Add missing header information to the base feed

1Maxnet1 commented:
I successfully created a setup including motis and realtime-metrics. It seems that our trip Updates are not yet in a state, where motis or gtfsrdb does like them. However we do successfully calculate some TripUpdates. I could not yet investigate in their quality, since they did not end in the database. So three known issues exist:

1. [ ] Sometimes the Trip-IDs used in the GTFS datasets do not match the IDs used in GTFS-RT feeds and the other way around. I wouldn't mind this issue too much, because it's a data issue, but this renders some feeds useless.
2. [ ] MOTIS and GTFSRD do not like/consume the resulting trip-updates, because they lack some information. We should add this lacking information and potentially need to retrieve it from the timetable or relatime inputs.
3. [ ] There appear a lot of warnings for stop-indexes, that are not valid. I just catch the exception and print a warning, but this should be further investigated.

For 1 I don't think it's feasible to come up with a real solution. I would just document it and it logs the issue but continues (as it already does to std:out, but maybe a log would be nice? -> separate Issue)
For 2 and 3: I would tackle this in extra issues, to not make the PR even bigger. If that's fine for you, please approve/review the current state. This would help me to continue my work :)

WuestFlorian reviewed (approved):
The implementation looks good to me, but I also didn't investgate whether or not the trip updates are of good quality. 

I agree that the observed problems (missing information in the feed and invalid stop-indexes) should be tackled in extra issues.

WuestFlorian commented:
You could add some more comments to clarify what certain sections are doing, to make it easier for other reviewers and to work on the code later on :)

1Maxnet1 committed:
Update variable names and add documentation

1Maxnet1 commented:
> You could add some more comments to clarify what certain sections are doing, to make it easier for other reviewers and to work on the code later on :)

Sure, added some documentation and renamed the variables to make it more clear what goes on. Did this help and are there any parts where you lack proper documentation/understanding?

1Maxnet1 commented:
Also created the follow-up issues:
- #27 
- #26 
- #25

WuestFlorian commented:
> > You could add some more comments to clarify what certain sections are doing, to make it easier for other reviewers and to work on the code later on :)
> 
> Sure, added some documentation and renamed the variables to make it more clear what goes on. Did this help and are there any parts where you lack proper documentation/understanding?

The additional documentation is good and especially the renamed variables are very helpful.
